File: src/domain/obsSetup.ts
/**
 * What putting a deal into OBS did, as `obs_setup.Result.to_json()` in Python
 * reports it, and the sentence that tells the creator. The work is the
 * server's; this only parses the answer and says it plainly, because a tool
 * that changes someone's scenes owes them an exact account.
 */

export type SetupAction = 'create' | 'repoint' | 'keep' | 'blocked';

export interface SetupStep {
  /** The OBS source, by its conventional name. */
  source: string;
  action: SetupAction;
  /** The deal it was showing before, when it was pointed elsewhere. */
  was: string | null;
}

export interface ObsSetupResult {
  deal: string;
  /** The scene new sources were added to: the one on screen at the time. */
  scene: string;
  steps: SetupStep[];
}

const ACTIONS: readonly SetupAction[] = ['create', 'repoint', 'keep', 'blocked'];

function parseStep(value: unknown): SetupStep | null {
  if (typeof value !== 'object' || value === null) return null;
  const o = value as Record<string, unknown>; // narrowed field by field below
  const action = ACTIONS.find((a) => a === o.action);
  if (typeof o.source !== 'string' || !action) return null;
  if (o.was !== null && typeof o.was !== 'string') return null;
  return { source: o.source, action, was: o.was };
}

/** The result as typed data, or null when the answer is not one. */
export function parseObsSetup(data: unknown): ObsSetupResult | null {
  if (typeof data !== 'object' || data === null) return null;
  const o = data as Record<string, unknown>; // narrowed field by field below
  if (typeof o.deal !== 'string' || typeof o.scene !== 'string' || !Array.isArray(o.steps)) return null;
  const steps = (o.steps as unknown[]).map(parseStep);
  return steps.every((s): s is SetupStep => s !== null) ? { deal: o.deal, scene: o.scene, steps } : null;
}

const list = (names: string[]): string =>
  names.length <= 1 ? names.join('') : `${names.slice(0, -1).join(', ')} and ${names[names.length - 1] ?? ''}`;

/** Exactly what was done to the creator's OBS, and what was left alone. */
export function setupSentence(result: ObsSetupResult): string {
  const named = (action: SetupAction) => result.steps.filter((s) => s.action === action);
  const parts: string[] = [];
  const made = named('create').map((s) => s.source);
  if (made.length > 0) {
    parts.push(`made ${list(made)} in the scene "${result.scene}", hidden until you show ${made.length === 1 ? 'it' : 'them'}`);
  }
  for (const step of named('repoint')) {
    parts.push(`pointed ${step.source} at this deal${step.was ? ` (it was showing ${step.was})` : ''}`);
  }
  const kept = named('keep').map((s) => s.source);
  if (kept.length > 0) parts.push(`${list(kept)} ${kept.length === 1 ? 'was' : 'were'} already right`);
  const blocked = named('blocked').map((s) => s.source);
  if (blocked.length > 0) {
    parts.push(`left ${list(blocked)} alone, because OBS has something else by that name that is not a browser source`);
  }
  const text = parts.join('; ');
  return text ? `${text.charAt(0).toUpperCase()}${text.slice(1)}.` : 'Nothing needed doing.';
}

File: python/sponsifer/obs_setup.py
"""Put a deal's placements into OBS, so nothing is copied and pasted by hand.

A streamer builds their scenes once: the emblem in the gameplay scene, the card
in the break scene. So this does not make a new set of sources per sponsor. It
creates the conventional sources that are missing, and points the ones that
exist at this deal. The scene layout stays the streamer's; the sponsor is what
changes.

What it will and will not do:

- **Only on an explicit request**, and only to the four conventional sources.
- **It creates hidden.** A new source goes into the scene that is on screen,
  switched off, at the canvas size. Nothing appears on stream that the streamer
  did not show.
- **It never removes, shows, hides or moves anything**, and it leaves alone a
  source whose address is already this server's, for this deal and this kind.
  An address that shows the right deal from somewhere else is still re-pointed:
  the first real OBS this met had a source aimed at a port nothing served any more.
- **Something else wearing a conventional name is never touched.** An image
  called "Sponsor overlay" is reported as in the way, not overwritten.
- **Never while the logger runs.** The log has already written down what each
  source was showing; `runner.with_obs` refuses, and this module never connects
  on its own.
- **The address is built here, from the server's own port**, never taken from a
  request, so the app cannot be used to make OBS load a page of someone's choosing.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal
from urllib.parse import parse_qs, urlsplit

from .obs import Obs
from .onair import PLACEMENT_KINDS, deal_shown

BROWSER_SOURCE = "browser_source"

Action = Literal["create", "repoint", "keep", "blocked"]


def overlay_url(port: int, deal_id: str, kind: str | None) -> str:
    """The overlay's address, as `overlayUrl` in src/domain/overlay.ts builds it for the same server."""
    return f"http://127.0.0.1:{port}/overlay.html?deal={deal_id}" + (f"&kind={kind}" if kind else "")


@dataclass(frozen=True, slots=True)
class Step:
    """What was done about one placement's source."""

    source: str
    action: Action
    #: The deal it was showing before, when it was pointed elsewhere.
    was: str | None = None


@dataclass(frozen=True, slots=True)
class Result:
    deal: str
    #: The scene new sources were added to: the one on screen at the time.
    scene: str
    steps: tuple[Step, ...]

    def to_json(self) -> dict[str, Any]:
        return {"deal": self.deal, "scene": self.scene,
                "steps": [{"source": s.source, "action": s.action, "was": s.was} for s in self.steps]}


def _already_right(url: str | None, port: int, deal_id: str, kind: str | None) -> bool:
    """True when the address is this server's own, for this deal and this kind. `localhost` is this server too."""
    if not url:
        return False
    parts = urlsplit(url)
    query = parse_qs(parts.query)
    return (parts.hostname in ("127.0.0.1", "localhost") and parts.port == port and parts.path == "/overlay.html"
            and query.get("deal", [""])[0] == deal_id and (query.get("kind", [None])[0] or None) == kind)


def _decide(obs: Obs, source: str, kind_in_obs: str | None, port: int, deal_id: str) -> Step:
    """Read-only: what this source needs."""
    if kind_in_obs is None:
        return Step(source, "create")
    if kind_in_obs != BROWSER_SOURCE:
        return Step(source, "blocked")
    url = obs.request("GetInputSettings", {"inputName": source}).get("inputSettings", {}).get("url")
    if _already_right(url, port, deal_id, PLACEMENT_KINDS[source]):
        return Step(source, "keep")
    showing = deal_shown(url)
    return Step(source, "repoint", was=showing if showing != deal_id else None)


def setup(obs: Obs, port: int, deal_id: str) -> Result:
    """Make OBS show this deal in every conventional source, and say what that took."""
    inputs = obs.request("GetInputList").get("inputs") or []
    kinds = {item.get("inputName"): item.get("unversionedInputKind") or item.get("inputKind") for item in inputs}
    scene = str(obs.request("GetCurrentProgramScene").get("currentProgramSceneName", ""))
    steps = tuple(_decide(obs, source, kinds.get(source), port, deal_id) for source in PLACEMENT_KINDS)

    video: dict[str, Any] | None = None
    for step in steps:
        url = overlay_url(port, deal_id, PLACEMENT_KINDS[step.source])
        if step.action == "repoint":
            obs.request("SetInputSettings", {"inputName": step.source, "inputSettings": {"url": url}, "overlay": True})
        elif step.action == "create":
            video = video or obs.request("GetVideoSettings")
            settings = {"url": url, "width": int(video.get("baseWidth", 1920)), "height": int(video.get("baseHeight", 1080))}
            obs.request("CreateInput", {"sceneName": scene, "inputName": step.source, "inputKind": BROWSER_SOURCE,
                                        "inputSettings": settings, "sceneItemEnabled": False})
    return Result(deal_id, scene, steps)
